pyop2/fusion/extended.py

```diff
@@ -0,0 +1,706 @@
+        """Initialize a :class:`fusion.Kernel` object.
+
+        :arg kernels: an iterator of some :class:`Kernel` objects. The objects
+            can be of class `fusion.Kernel` or of any superclass.
+        :arg fused_ast: the abstract syntax tree of the fused kernel. If not
+            provided, objects in ``kernels`` are considered "isolated C functions".
+        :arg loop_chain_index: index (i.e., position) of the kernel in a loop chain.
+            Meaningful only if ``fused_ast`` is specified.
+        """
+        # Protect against re-initialization when retrieved from cache
+        if self._initialized:
+            return
+        Kernel._globalcount += 1
+
+        # We need to distinguish between the kernel name and the function name(s).
+        # Since /fusion.Kernel/ are, in general, collections of functions, the same
+        # function (which is itself associated a Kernel) can appear in different
+        # /fusion.Kernel/ objects, but possibly under a different name (to avoid
+        # name clashes)
+        self._name = "_".join([k.name for k in kernels])
+        self._function_names = {self.cache_key: self._name}
+
+        self._cpp = any(k._cpp for k in kernels)
+        self._opts = dict(flatten([k._opts.items() for k in kernels]))
+        self._include_dirs = list(set(flatten([k._include_dirs for k in kernels])))
+        self._ldargs = list(set(flatten([k._ldargs for k in kernels])))
+        self._headers = list(set(flatten([k._headers for k in kernels])))
+        self._user_code = "\n".join(list(set([k._user_code for k in kernels])))
+        self._attached_info = {'fundecl': None, 'attached': False}
+
+        # What sort of Kernel do I have?
+        if fused_ast:
+            # A single AST (as a result of soft or hard fusion)
+            self._ast = fused_ast
+            self._code = self._ast_to_c(fused_ast)
+        else:
+            # Multiple functions (AST or strings, as a result of tiling)
+            self._ast = None
+            self._code = self._multiple_ast_to_c(kernels)
+        self._kernels = kernels
+
+        self._initialized = True
+
+    def __iter__(self):
+        for k in self._kernels:
+            yield k
+
+    def __str__(self):
+        return "OP2 FusionKernel: %s" % self._name
+
+
+# API for fused parallel loops
+
+class ParLoop(sequential.ParLoop):
+
+    """The root class of non-sequential parallel loops."""
+
+    pass
+
+
+class FusionParLoop(ParLoop):
+
+    def __init__(self, kernel, iterset, *args, **kwargs):
+        self._it_space = kwargs['it_space']
+        super(FusionParLoop, self).__init__(kernel, iterset, *args, **kwargs)
+
+    def _build_itspace(self, iterset):
+        """
+        Bypass the construction of a new iteration space.
+
+        This avoids type checking in base.ParLoop._build_itspace, which would
+        return an error when the fused loop accesses arguments that are not
+        accessed by the base loop.
```

Comment:
When I fuse two kernels over different iteration spaces (e.g., in HardFusion), say X (base) and Y (fused) it may be that Y accesses fields that are not accessed by X. Now, Z = fuse(X, Y) has the same iteration space as X (base), but it might also have an indirect ``Arg`` from Y's iteration space to another set. In ``base.ParLoop``'s build_itspace we do type checking: we check that all maps from all indirect arguments have as ``fromset`` exactly X's iteration space -- which wouldn't be the case anymore here. 

Roughly speaking, a ``FusionParLoop`` relaxes the constraint that all indirect arguments should have the same base set (i.e., the iteration set of the loop)